### PR TITLE
Fixed detailed balance and sampling issues with shape_move.ShapeSpace

### PR DESCRIPTION
**Pull request by janbridley**
## Description

- Removed scaling for shape moves near extrema in shape space.
- Added mirror boundary on shape axes to improve sampling near extrema in shape space.

## Motivation and context

Previously, the shape move step size would change when approaching the extrema of each shape space axis (0 and 1). This behavior has been removed to ensure detailed balance is maintained. Furthermore, a mirror boundary has been implemented on each shape space axis. Proposed shape moves that would otherwise provide an invalid value (and therefore be rejected) will instead "bounce" off the end of the shape axis, resulting in a smaller shape move. Each of these boundary-affected moves is paired with an inverse move in the opposite direction, preserving detailed balance.

## How has this been tested?

All pytests run as expected, and HOOMD can be built properly with these changes.

## Change log

- Removed scaling for shape moves near a boundary in shape space.
- Added mirror boundary on shape axes to improve sampling near boundaries in shape space.

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/hpmc/ShapeMoves.h [hunk 1]
```diff
@@ -146,18 +146,30 @@ template<typename Shape> class PythonShapeMove : public ShapeMoveBase<Shape>
...
             Scalar x = (r < this->m_move_probability) ? uniform(rng) : 0.0;
-            m_params[type_id][i] += x;
+            // Reflect trial moves about boundaries
+            if (m_params[type_id][i] + x > 1)
+                {
+                m_params[type_id][i] = 2 - (m_params[type_id][i] + x);
+                }
+            else if (m_params[type_id][i] + x < 0)
+                {
+                m_params[type_id][i] = x - m_params[type_id][i];
```
Comment (cbkerr) [inline hoomd/hpmc/ShapeMoves.h]: Jen and I discussed offline and we think we need this math to reflect across 0:

```suggestion
                m_params[type_id][i] = -(m_params[type_id][i] + x);
```

File: hoomd/hpmc/ShapeMoves.h [hunk 2]
```diff
@@ -146,18 +146,30 @@ template<typename Shape> class PythonShapeMove : public ShapeMoveBase<Shape>
...
+                }
+            else
+                {
+                m_params[type_id][i] += x;
+                }
+            pybind11::object d = m_python_callback(type_id, m_params[type_id]);
+            pybind11::dict shape_dict = pybind11::cast<pybind11::dict>(d);
+            shape = typename Shape::param_type(shape_dict);
             }
```
Comment (cbkerr) [inline hoomd/hpmc/ShapeMoves.h]: I don't think this needs to be in the for loop:
```suggestion
            }
        pybind11::object d = m_python_callback(type_id, m_params[type_id]);
        pybind11::dict shape_dict = pybind11::cast<pybind11::dict>(d);
        shape = typename Shape::param_type(shape_dict);
```

File: hoomd/hpmc/ShapeMoves.h (hunk 1 above)
Comment (janbridley) [reply hoomd/hpmc/ShapeMoves.h]: This is correct - the initial code made the assumption x > 0, which is not true for proposed moves that would bring m_params[type_id] < 0

File: hoomd/hpmc/ShapeMoves.h [hunk 3]
```diff
@@ -146,18 +146,30 @@ template<typename Shape> class PythonShapeMove : public ShapeMoveBase<Shape>
         for (unsigned int i = 0; i < m_params[type_id].size(); i++)
             {
             Scalar stepsize = this->m_step_size[type_id];
-            Scalar a = fmax(-stepsize, -(m_params[type_id][i]));
-            Scalar b = fmin(stepsize, (1.0 - m_params[type_id][i]));
+            Scalar a = -stepsize;
+            Scalar b = stepsize;
+
             hoomd::UniformDistribution<Scalar> uniform(a, b);
```
Comment (joaander) [inline hoomd/hpmc/ShapeMoves.h]:
```suggestion
            hoomd::UniformDistribution<Scalar> uniform(-stepsize, stepsize);
```

File: hoomd/hpmc/ShapeMoves.h [hunk 4]
```diff
@@ -146,18 +146,30 @@ template<typename Shape> class PythonShapeMove : public ShapeMoveBase<Shape>
...
+            else
+                {
+                m_params[type_id][i] += x;
+                }
             }
         pybind11::object d = m_python_callback(type_id, m_params[type_id]);
         pybind11::dict shape_dict = pybind11::cast<pybind11::dict>(d);
         shape = typename Shape::param_type(shape_dict);
         }
-
```
Comment (joaander) [inline hoomd/hpmc/ShapeMoves.h]: There should be a blank line here for readability. I'm not sure why clang-format doesn't add it.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! Two small style changes and this is good to merge.

File: hoomd/hpmc/ShapeMoves.h (hunk 4 above)
Comment (janbridley) [reply hoomd/hpmc/ShapeMoves.h:160]:
```suggestion
            }
        pybind11::object d = m_python_callback(type_id, m_params[type_id]);
        pybind11::dict shape_dict = pybind11::cast<pybind11::dict>(d);
        shape = typename Shape::param_type(shape_dict);
        }
        
```

Comment (janbridley) [reply hoomd/hpmc/ShapeMoves.h:160]:
```suggestion
            }
        pybind11::object d = m_python_callback(type_id, m_params[type_id]);
        pybind11::dict shape_dict = pybind11::cast<pybind11::dict>(d);
        shape = typename Shape::param_type(shape_dict);
        }
        
```

**Review by cbkerr — APPROVED** [review]

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**